### DISCUSSION TITLE
Non-owning side of 1:1 relationship not persisted?

### DISCUSSION THREAD

**Discussion by johanneslumpe** [mikro-orm/mikro-orm#4595] [Q&A] [answered]
I have two models, pared down for brevity:

```ts
@Entity()
export class User extends BaseEntity  {
  @Property({ unique: true })
  username!: string;

  @Property({ unique: true })
  email!: string;

  @Property({ type: 'timestamp', nullable: true })
  termsAcceptedAt?: Date | null;

  @Property({ version: true })
  version!: number;

  @Enum({ items: () => UserRole, array: true, default: [UserRole.USER] })
  roles: UserRole[] = [UserRole.USER];

  @OneToOne(() => UserProfile, { inversedBy: 'user', ref: true, nullable: true })
  userProfile?: Ref<UserProfile>;
}


@Entity()
export class UserProfile extends BaseEntity {
  @Property()
  firstName!: string;

  @Property()
  lastName!: string;

  @Property({ columnType: 'text', default: '' })
  aboutMe = '';

  @Property({ version: true })
  version!: number;

  @OneToOne(() => User, { mappedBy: 'userProfile', ref: true })
  user!: Ref<User>;
}
```

Everything so far has been working out great. Now I'm trying to create a new user profile in my service. The whole API is behind auth, so during the request flow I'm loading the current user via the repository and set it on my request context.

In my GQL resolver I'm calling my service to create a profile:
```ts
await this.userProfileService.createProfile({
      ...profileData,
      user: context.currentUserReference,
});
await context.flushDatabaseChanges(); // this just calls the current em's `flush` method.
```

Here's the body of that method:
```ts
public async createProfile(
    profileData: Omit<RequiredEntityData<UserProfile>, 'user'> & { user: Ref<User> },
  ): Promise<UserProfile> {
    const existingProfile = await this.userProfileRepository.findByUserId(profileData.user.id);
    if (existingProfile) {
      throw new Error(`A profile already exists for user with id "${profileData.user.id}"`);
    }
    const userProfile = this.userProfileRepository.create({
      ...profileData,
    });
    return userProfile;
  }
```

The user which the profile is attached to does not get updated, or at least the relationship property isn't updated. ~I also tried to make arbitrary changes to that user object in my context, e.g. username or email. Nothing is persisted in this call~. 

_Edit: updating a user's username works fine in this context. It is only the relational property that does not get persisted, not matter how I try to set or save it. Event setting the `userProfile` property on the user object direct with a reference does not trigger an update._

What is very odd is that the version of the `user` entity does get updated when trying to save the user profile relation.

Am I missing something obvious?

**Comment by johanneslumpe** [comment]
@B4nan Hoping that you have some grand insight here :)

**Comment by B4nan** [comment]
Are you sure the `User` entity from your context is managed by the same EM? 

There is one (yet undocumented) debugging feature you can try - the `MIKRO_ORM_LOG_EM_ID` env var, once you set it to `true` it will start printing the EM id when you `console.log(entity)`.

Also, try to do the update on `User` side directly or via `assign` instead of `em.create`, if that works, I might have one idea. First, create the profile entity without the user ref, and then set it to the `User.userProfile`.

**Reply by johanneslumpe** [reply]
So I checked both entities with regards to which entity manager is managing them. The user entity is managed by instance `1` and the user profile is showing up as not being managed, even though `persistOnCreate` is enabled. I also tried to mark the user profile model as persisted explicitly:
```ts
public create(data: RequiredEntityData<UserProfile>): UserProfile {
    const model = this.#ormRepository.create(data);
    this.#entityManager.persist(model);
    return model;
  }
```;

this did not change anything. It's still logging as not being managed. I assume this is expected behavior for newly created entities as this seems to be the case for other parts of the app as well, whenever a new entity is created.

I tried assigning the user profile directly to the user instance, without setting `user` on the profile using
```ts
Object.assign(context.user, { userProfile: profile });
```

This also didn't cause the user to update. I double checked the id of the em I'm flushing, and that is indeed the em with id `1`.

I should also mention that I am running my tests using 
```
allowGlobalContext: true,
implicitTransactions: false,
```

But I am observing the same behavior when executing this through the actual api, outside of tests. The EM id properly increases for each request, each repo is instantiated once, but is using an injected entity manager, which in turn is using a request scoped context. If i log out the em ids in my tests it's obviously always id `1`, but logging the id of the entity manager that's being flushed and used to resolve repositories when running the actual api properly changes with every request.

**Reply by johanneslumpe** [reply]
@B4nan I just found a way to make this work:

```ts
    const profile = await this.userProfileService.createProfile({
      ...profileData,
      user: context.currentUserReference,
    });
    await context.flushDatabaseChanges();
    Object.assign(context.user, { userProfile: profile });
    await context.flushDatabaseChanges();
```

If I flush the changes after creating the profile (without assigning the user) and then I assign the profile to the user and flush again, it gets properly persisted. That seems rather odd to me, as I'd expect to only flush once at the end of my mutation. If I flush right after profile creation though I can see that the profile is managed by the correct entity manager after that happened.

 If I have the user already assigned to the profile then the above code does not work anymore.

Whether I'm using the actual profile or a reference does not seem to matter either, both work just fine if assigned to the user _after_ the profile has been created.

What doesn't work is this:
```ts
 const profile = await this.userProfileService.createProfile({
      ...profileData,
      user: context.currentUserReference,
    });

    await context.flushDatabaseChanges();

    profile.user = context.currentUserReference;
    await context.flushDatabaseChanges();
```

Here I create the profile without a user reference, flush it, then assign the current user and persist again. In this case, the relation is _not_ updated. No matter whether I use a ref or the full entity instance. In this case, no queries are fired at all to update the user object. 

It sure feels like there is something funky going on with reverse-assigning of a 1:1 relationship.

**Reply by B4nan** [reply]
> The user entity is managed by instance 1

Without even reading the rest, this is _very_ suspicious, it means the entity is managed by the global EM - you should never use that one, and unless you use `allowGlobalContext` option (which you should pretty much _never_ do outside of tests), you would get a validation error if you try doing that.

edit: right, you are talking about tests - so those are some unit tests with mocks, where you have no sense of request context?

**Reply by B4nan** [reply]
I will need to see a complete but minimal repro, and by minimal I mean no webserver and GQL. The end goal should be to minimize your repro so it is using one EM instance instead of repositories and services, to make it clear how the code flows and scratch out possible context mishandling.

> the user profile is showing up as not being managed

Yes, this is expected, technically it will become managed after the flush call resolves, not after the `persist` call.

> though persistOnCreate is enabled

FYI it is enabled by default nowadays.

>     Object.assign(context.user, { userProfile: profile });

Keep in mind this is exactly the same as `context.user.userProfile = profile`, `Object.assign` here is not doing any magic. Even like this, it will trigger propagation (setting `userProfile.user = user`). But this brings me to one conclusion - you are mapping both of those relations to use the `Ref` wrapper, so assigning the naked entity to them is wrong, it needs to be the `Reference` instance instead. Maybe the problem will be in the Ref wrapper handling (either yours or internal).

> context.currentUserReference

What is actually in this variable? Is it a loaded User entity, a User entity reference (so not initialized User entity), or a reference wrapper (so the `Reference<User>` class, wrapping the entity)? Where and how are you creating that?

**Comment by johanneslumpe** [comment]
> edit: right, you are talking about tests - so those are some unit tests with mocks, where you have no sense of request context?

In tests no request context is created, correct. I am under the impression that if I created a request context, I wouldn't be able to control the overall transaction within my tests and rollback all DB writes once the test concludes. I also set up test data directly within my tests, which are outside of the api's request context, so those values also wouldn't be readable from within a new request context, as the transaction itself isn't being committed.

> Keep in mind this is exactly the same as context.user.userProfile = profile, Object.assign here is not doing any magic. Even like this, it will trigger propagation (setting userProfile.user = user). But this brings me to one conclusion - you are mapping both of those relations to use the Ref wrapper, so assigning the naked entity to them is wrong, it needs to be the Reference instance instead. Maybe the problem will be in the Ref wrapper handling (either yours or internal).

Yeah, I was just documenting both approaches for clarity, aware that `Object.assign` doesn't do anything special. Regarding having to use a ref: correct, it needs to be a ref. I tried it both with a ref and with the profile, just to be exhaustive. Both work as expected with the double flush.

> What is actually in this variable? Is it a loaded User entity, a User entity reference (so not initialized User entity), or a reference wrapper (so the Reference<User> class, wrapping the entity)? Where and how are you creating that?

Here is the snippet from my JWT auth handler that looks up a user within a current request context then assigns it:
```ts
   const userRepository = request.diContainer.get<UserRepository>(TOKENS.userRepository);
   const user = await userRepository.findByFusionAuthId(decodedUser.fusionAuthId);
   request.requestContext.set('user', user);
   request.requestContext.set('userReference', userRepository.getPrimaryKeyReference(user.id));
```

I did a quick test by hacking in a code path that doesn't use request context nor any repositories:
```ts
const [user] = await orm.em.find(User, {});
const userProfile = orm.em.create(UserProfile, {
        firstName: 'testing',
        lastName: 'testing',
        user: Reference.create(user),
      });
await orm.em.flush();
```

I am observing the same behavior here.

I have pushed an isolated reproduction to this repo: https://github.com/johanneslumpe/mikroorm-ref-bug-repro

**Reply by B4nan** [reply]
It seems to be working fine, at least when extracted and executed in the ORM test suite:

```ts
import { Entity, PrimaryKey, Property, OptionalProps, OneToOne, Ref } from '@mikro-orm/core';
import { MikroORM } from '@mikro-orm/sqlite';

import { v4 } from 'uuid';

export abstract class BaseEntity<Optional = never> {

  [OptionalProps]?: Optional | 'createdAt' | 'updatedAt';

  @PrimaryKey({ columnType: 'uuid' })
  id: string = v4();

  @Property()
  createdAt: Date = new Date();

  @Property({ onUpdate: () => new Date() })
  updatedAt: Date = new Date();

}

@Entity()
export class User extends BaseEntity<'version'> {

  @Property({ unique: true })
  username!: string;

  @Property({ unique: true })
  email!: string;

  @Property({ version: true })
  version!: number;

  @OneToOne(() => UserProfile, {
    inversedBy: 'user',
    ref: true,
    nullable: true,
  })
  userProfile?: Ref<UserProfile>;

}

@Entity()
export class UserProfile extends BaseEntity<'version'> {

  @Property()
  firstName!: string;

  @Property()
  lastName!: string;

  @Property({ version: true })
  version!: number;

  @OneToOne(() => User, { mappedBy: 'userProfile', ref: true })
  user!: Ref<User>;

}

let orm: MikroORM;

beforeAll(async () => {
  orm = await MikroORM.init({
    entities: [User],
    dbName: ':memory:',
    debug: true,
  });
  await orm.schema.createSchema();
});

afterAll(async () => {
  await orm.close(true);
});

test(`GH issue 4595`, async () => {
  const em = orm.em.fork();
  em.create(User, {
    email: 'example@example.com',
    username: 'testing',
  });
  await em.flush();

  const em2 = orm.em.fork();
  const [user] = await em2.find(User, {});

  const userProfile = em2.create(UserProfile, {
    firstName: 'testing',
    lastName: 'testing',
    user,
  });
  await em2.flush();

  // use a fresh em to find the user again with their profile populated
  const em3 = orm.em.fork();
  const userWithProfilePopulated = await em3.findOne(
    User,
    { id: user.id },
    { populate: ['userProfile'] },
  );

  console.log({ userWithProfilePopulated });

  // alternate way of querying for the user just to show that the column really isn't set, as no user will be found
  const userQueriedByProfileColumn = await em3.findOne(
    User,
    { id: user.id, userProfile },
    { populate: true },
  );

  console.log({ userQueriedByProfileColumn });
});
```

Produces following logs:

```
  console.log
    [query] begin

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] insert into `user` (`id`, `created_at`, `updated_at`, `username`, `email`) values ('47ace310-a902-4373-8788-cbfa5093985d', 1691609286471, 1691609286471, 'testing', 'example@example.com') returning `id`, `version` [took 0 ms]

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] commit

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] select `u0`.* from `user` as `u0` [took 0 ms, 1 result]

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] begin

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] insert into `user_profile` (`id`, `created_at`, `updated_at`, `first_name`, `last_name`) values ('d0ab95b5-63c9-4593-b847-804e6aef99fd', 1691609286506, 1691609286506, 'testing', 'testing') returning `id`, `version` [took 0 ms]

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] update `user` set `user_profile_id` = 'd0ab95b5-63c9-4593-b847-804e6aef99fd', `updated_at` = 1691609286507, `version` = `version` + 1 where `id` = '47ace310-a902-4373-8788-cbfa5093985d' and `version` = 1 [took 1 ms]

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] select `u0`.`id`, `u0`.`version` from `user` as `u0` where `u0`.`id` in ('47ace310-a902-4373-8788-cbfa5093985d') [took 0 ms, 1 result]

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] commit

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] select `u0`.* from `user` as `u0` where `u0`.`id` = '47ace310-a902-4373-8788-cbfa5093985d' limit 1 [took 0 ms, 1 result]

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    [query] select `u0`.*, `u1`.`id` as `user_id` from `user_profile` as `u0` left join `user` as `u1` on `u0`.`id` = `u1`.`user_profile_id` where `u0`.`id` in ('d0ab95b5-63c9-4593-b847-804e6aef99fd') order by `u0`.`id` asc [took 0 ms, 1 result]

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    {
      userWithProfilePopulated: User {
        id: '47ace310-a902-4373-8788-cbfa5093985d',
        createdAt: 2023-08-09T19:28:06.471Z,
        updatedAt: 2023-08-09T19:28:06.507Z,
        username: 'testing',
        email: 'example@example.com',
        version: 2,
        userProfile: Ref<UserProfile<d0ab95b5-63c9-4593-b847-804e6aef99fd>> { entity: [UserProfile] }
      }
    }

      at Object.<anonymous> (tests/issues/GH1003.test.ts:100:11)

  console.log
    [query] select `u0`.* from `user` as `u0` where `u0`.`id` = '47ace310-a902-4373-8788-cbfa5093985d' and `u0`.`user_profile_id` = 'd0ab95b5-63c9-4593-b847-804e6aef99fd' limit 1 [took 0 ms, 1 result]

      at DefaultLogger.log (packages/core/src/logging/DefaultLogger.ts:34:10)

  console.log
    {
      userQueriedByProfileColumn: User {
        id: '47ace310-a902-4373-8788-cbfa5093985d',
        createdAt: 2023-08-09T19:28:06.471Z,
        updatedAt: 2023-08-09T19:28:06.507Z,
        username: 'testing',
        email: 'example@example.com',
        version: 2,
        userProfile: Ref<UserProfile<d0ab95b5-63c9-4593-b847-804e6aef99fd>> { entity: [UserProfile] }
      }
    }

      at Object.<anonymous> (tests/issues/GH1003.test.ts:109:11)
```

**Reply by B4nan** [reply]
Oh I think I got it - you are targeting `esnext`:

https://github.com/johanneslumpe/mikroorm-ref-bug-repro/blob/master/tsconfig.json#L4C1-L4C24

Which means `"useDefineForClassFields": true` and that means broken propagation. Try to switch to `"useDefineForClassFields": false` that should help. See #4216 (and #4434 with a fix for v6). 

---

I should really find a way to detect this flag and print some warning in v5...

**Reply by johanneslumpe** [reply]
@B4nan That was sadly not the issue. I have disable the option and I get the same behavior. Looking at your test file I noticed that it runs against sqlite and that got me thinking. I set up a new test that compares the sqlite package with the postgres package.

It runs the same code that is run within your test suite, but it'll run it once against sqlite, and then against postgres. Here is the logged output:

```
==============
==============
TESTING: sqlite
==============
==============
[query] select `u0`.* from `user` as `u0` [took 1 ms, 1 result]
[query] begin
[query] insert into `user_profile` (`id`, `created_at`, `updated_at`, `first_name`, `last_name`) values ('197ac8d0-f93f-48c6-ace4-2718fa1a3081', 1691617349577, 1691617349577, 'testing', 'testing') returning `id`, `version` [took 1 ms]
[query] update `user` set `user_profile_id` = '197ac8d0-f93f-48c6-ace4-2718fa1a3081', `updated_at` = 1691617349579, `version` = `version` + 1 where `id` = '78f2d720-6b3f-4f00-b49c-ffa3ffaaeefc' and `version` = 1 [took 1 ms]
[query] select `u0`.`id`, `u0`.`version` from `user` as `u0` where `u0`.`id` in ('78f2d720-6b3f-4f00-b49c-ffa3ffaaeefc') [took 1 ms, 1 result]
[query] commit
[query] select `u0`.* from `user` as `u0` where `u0`.`id` = '78f2d720-6b3f-4f00-b49c-ffa3ffaaeefc' limit 1 [took 0 ms, 1 result]
[query] select `u0`.*, `u1`.`id` as `user_id` from `user_profile` as `u0` left join `user` as `u1` on `u0`.`id` = `u1`.`user_profile_id` where `u0`.`id` in ('197ac8d0-f93f-48c6-ace4-2718fa1a3081') order by `u0`.`id` asc [took 1 ms, 1 result]
{
  userWithProfilePopulated: User [managed by 6] {
    id: '78f2d720-6b3f-4f00-b49c-ffa3ffaaeefc',
    createdAt: 2023-08-09T21:42:29.553Z,
    updatedAt: 2023-08-09T21:42:29.579Z,
    username: 'testing',
    email: 'example@example.com',
    version: 2,
    userProfile: Ref<UserProfile<197ac8d0-f93f-48c6-ace4-2718fa1a3081>> { entity: [UserProfile [managed by 6]] }
  }
}
[query] select `u0`.* from `user` as `u0` where `u0`.`id` = '78f2d720-6b3f-4f00-b49c-ffa3ffaaeefc' and `u0`.`user_profile_id` = '197ac8d0-f93f-48c6-ace4-2718fa1a3081' limit 1 [took 1 ms, 1 result]
{
  userQueriedByProfileColumn: User [managed by 6] {
    id: '78f2d720-6b3f-4f00-b49c-ffa3ffaaeefc',
    createdAt: 2023-08-09T21:42:29.553Z,
    updatedAt: 2023-08-09T21:42:29.579Z,
    username: 'testing',
    email: 'example@example.com',
    version: 2,
    userProfile: Ref<UserProfile<197ac8d0-f93f-48c6-ace4-2718fa1a3081>> { entity: [UserProfile [managed by 6]] }
  }
}
```

```
==============
==============
TESTING: postgres
==============
==============
[query] select "u0".* from "user" as "u0" [took 4 ms, 1 result]
[query] begin
[query] insert into "user_profile" ("id", "created_at", "updated_at", "first_name", "last_name") values ('3d6aecdf-a5b4-42e0-bc62-eb395fbbc6aa', '2023-08-09T21:57:40.966Z', '2023-08-09T21:57:40.966Z', 'testing', 'testing') returning "id", "version" [took 2 ms]
[query] update "user" set "updated_at" = '2023-08-09T21:57:40.967Z', "version" = "version" + 1 where "id" = '798ae006-5cac-48d6-b6fe-2936790fde9c' and "version" = 69 [took 2 ms]
[query] select "u0"."id", "u0"."version" from "user" as "u0" where "u0"."id" in ('798ae006-5cac-48d6-b6fe-2936790fde9c') [took 1 ms, 1 result]
[query] commit
[query] select "u0".* from "user" as "u0" where "u0"."id" = '798ae006-5cac-48d6-b6fe-2936790fde9c' limit 1 [took 2 ms, 1 result]
{
  userWithProfilePopulated: User [managed by 9] {
    id: '798ae006-5cac-48d6-b6fe-2936790fde9c',
    createdAt: 2023-08-09T13:41:45.000Z,
    updatedAt: 2023-08-09T21:57:41.000Z,
    username: 'testing',
    email: 'example@example.com',
    version: 70,
    userProfile: null
  }
}
[query] select "u0".* from "user" as "u0" where "u0"."id" = '798ae006-5cac-48d6-b6fe-2936790fde9c' and "u0"."user_profile_id" = '3d6aecdf-a5b4-42e0-bc62-eb395fbbc6aa' limit 1 [took 2 ms, 0 result]
{ userQueriedByProfileColumn: null }
```

This is using the exact same test code. The only difference is the adapter used. Sqlite is running in memory, so I also tried changing that to a local file and priming the user into it, to rule out a difference there. Sqlite works regardless of whether the db is in memory or in a local file. Postgres does not.

Whether `useDefineForClassFields` is `true` or `false` does not seem to have any impact on the results. The tests always pass for sqlite, always fail for postgres. 

Check out the updated repro and do run `npm run showcase-sqlite-vs-postgres` to see the comparison.

I feel like I must be missing something?!

**Reply by B4nan** [reply]
Ok, will test with postgres when I have some spare time, could be about that for sure. I thought you said it works the same with sqlite and pg, but I see I confused myself with some other bug report.

**Reply by B4nan** [reply]
The test I provided works fine with postgres too... 

I would really blame the tsconfig, usually things I cannot reproduce inside the ORM test suite are about that.

**Reply by johanneslumpe** [reply]
@B4nan just matched the tsconfig in the example repo to the one you are using in the repo, and it still fails. Did you observe the failure when you ran my example? Or did you run it only in the dedicated test suite? (Would be nice to know if you can reproduce it there as well, I feel like I'm crazy) I really can't see how this is related to the tsconfig now that it matches yours. I've tested it on both node 16 and 18 to ensure there wasn't any issue there. 

This is the current config:

```
{
  "compilerOptions": {
    "module": "Node16",
    "target": "es2020",
    "lib": ["es2020"],
    "moduleResolution": "Node16",
    "declaration": true,
    "strict": true,
    "suppressImplicitAnyIndexErrors": true,
    "esModuleInterop": true,
    "preserveConstEnums": true,
    "ignoreDeprecations": "5.0",
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "sourceMap": true,
    "outDir": "build",
    "rootDir": "./src"
  }
}
```

Setting or not setting `outDir` and `rootDir` is not causal. That should rule out any issues related to the config no? I'm rather confused that it consistently works in your test suite but consistently fails for me in a real-world example. That said, the fact that it works for sqlite would be indicative that there is something else going on as the expectation would be that sqlite would fail as well, if it was really about faulty compilation.

Do you have some pointers where I could debug further to aid in figuring this out?

**Reply by johanneslumpe** [reply]
@B4nan I just ran my repro code with the following versions:

```
@mikro-orm/cli@5.6.16
@mikro-orm/core@5.6.16 
@mikro-orm/migrations@5.6.16 
@mikro-orm/postgresql@5.6.16 
@mikro-orm/sqlite@5.6.16
```

And, to my surprise, the code works as expected now. Coincidentally this is the same version I was running before I updated to 5.7.x and reported that something had changed around using `{ value: undefined }` in a filter query. I also tested the 6.x release and it is exhibiting the same failure behavior as 5.7.x. Was there a significant change between 5.6 and 5.7?

**Reply by B4nan** [reply]
Interesting. I am testing on latest master, so v5.7.14 plus a few small changes, and it works fine in there. You could try latest dev version which is exactly the code from master.

Let me check your repro to see if that will fail for me or not.

**Reply by B4nan** [reply, marked as answer]
Your repro is working as well if I update the ORM to latest version - not even dev version, v5.7.14 works just fine, any reason why did you not upgrade first? :]

**Reply by johanneslumpe** [reply]
I could swear I tried that and it wasn’t working… let me try again. Out of curiosity: did the repro fail for you with other 5.7 versions?

**Reply by B4nan** [reply]
Sounds more like you have some packages on different versions (or you have multiple core packages installed), there shouldnt be any differences or type level breaking changes. Maybe try to wipe lock file and reinstall. 

And yes, the repro failed with 5.7.12, then i tried latest dev and latest stable, both worked, havent checked 5.7.13...

And no worries, happy that it worked in the end. It could be as well a combination of this plus the target: esnext, that can have very similar symptoms.

**Reply by johanneslumpe** [reply]
@B4nan Indeed - the repro works with the latest version. I'm really really flabbergasted. I'm so sorry for all this back and forth when upgrading, which I thought I tried first, would've just solved it.

**Reply by johanneslumpe** [reply]
@B4nan Yeah just removed my comment - that was all my fault: duplicate packages. I think everything's good now! Do ping me when you want to discuss potential filter query changes for version 6 :)

**Reply by B4nan** [reply]
It's weird that you didn't get the [validation warning](https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/ConfigurationLoader.ts#L281).

What filter query changes do you have on your mind? :]

**Reply by johanneslumpe** [reply]
I didn't even get to the warning because of the type errors I got prior to compiling the code (I don't run ts-node). I most likely would've gotten that error. The issue was one set of versions installed in a workspace, another one at the root. That somehow tripped TS up during type resolution it seems.

Regarding filters: I have queries similar to this in 5.6:

```ts
function ({some, other }: ({ some: string, other?: string}) {
    return repo.findOneOrFail({
      some,
      other,
    });
}
```
They worked just fine, by omitting `other` when it was `undefined` after upgrading to 5.7 those changed to produce queries that would contain `other IS NULL`, which I found rather confusing as I'd expect `IS NULL` to be a specific keyword not the default behavior. That said, given that we have the `forceUndefined` flag, I can see how users using it might expect the `IS NULL` query vs omitting `undefined` values. What are your thoughts on this?

**Reply by B4nan** [reply]
> after upgrading to 5.7

Oh, this issue, I thought you mean some type-level improvements, as `FilterQuery` is just a type. I'd still very much like to see a repro for that (I mean for the previous behavior that ignored such values, so for v5.6). But what I said in that issue still holds, it was always supposed to be working like that (for years, IIRC from the very beginning actually) and if there was some way to make it ignore the undefined keys, that was a bug that got fixed (probably as a side effect), not a new breaking change IMO (although any bug fix is in a way a breaking change).

We could lift this restriction in v6 or make this conditional, but there would have to be additional guards in place most probably - as we need to protect from doing `findOneOrFail({})` as a result of having all the values undefined, it would be quite a huge security hole. But maybe that case is already covered, I think calling `findOne({})` already fails.

Do you know if other ORMs behave this way?

**Reply by johanneslumpe** [reply]
I'll try to get a repro going, but I honestly wouldn't be surprised if it was some odd combination of conditions that caused it that simply made it look like it was working that way.

I don't remember off the top of my head if type-orm behaves this way, I'll have to check. That said, I don't think this is a huge deal if it stays the way it is.
